File: frontend/src/services/mock/dashboardService.ts
import { DashboardData, DashboardKpi, DashboardAlert, MonthlyTrendData } from '../../types';

export class DashboardService {
  // モックデータ: KPI
  private mockKpis: DashboardKpi[] = [
    {
      id: 'kpi-1',
      label: '総売上',
      value: 8500000,
      unit: '¥',
      comparison: {
        trend: 'positive',
        monthOverMonth: 5.2,
        yearOverYear: 12.3,
      },
    },
    {
      id: 'kpi-2',
      label: '営業利益',
      value: 2100000,
      unit: '¥',
      comparison: {
        trend: 'negative',
        monthOverMonth: 3.8,
        yearOverYear: -2.1,
      },
    },
    {
      id: 'kpi-3',
      label: '粗利率',
      value: 24.7,
      unit: '%',
      comparison: {
        trend: 'negative',
        monthOverMonth: -0.5,
        yearOverYear: -3.2,
      },
    },
    {
      id: 'kpi-4',
      label: '総患者数',
      value: 420,
      unit: '人',
      comparison: {
        trend: 'positive',
        monthOverMonth: 8,
        yearOverYear: 35,
      },
    },
    {
      id: 'kpi-5',
      label: 'ユニット稼働率',
      value: 78.5,
      unit: '%',
      comparison: {
        trend: 'positive',
        monthOverMonth: 2.3,
        yearOverYear: 5.7,
      },
    },
    {
      id: 'kpi-6',
      label: '自費率',
      value: 18.2,
      unit: '%',
      comparison: {
        trend: 'negative',
        monthOverMonth: -1.2,
        yearOverYear: -0.3,
      },
    },
    {
      id: 'kpi-7',
      label: '新患数',
      value: 35,
      unit: '人',
      comparison: {
        trend: 'positive',
        monthOverMonth: 5,
        yearOverYear: 8,
      },
    },
    {
      id: 'kpi-8',
      label: 'リコール率',
      value: 62.3,
      unit: '%',
      comparison: {
        trend: 'positive',
        monthOverMonth: -0.2,
        yearOverYear: 3.5,
      },
    },
  ];

  // モックデータ: アラート
  private mockAlerts: DashboardAlert[] = [
    {
      id: 'alert-1',
      severity: 'warning',
      title: '自費率が低下しています',
      message: '前月比 -1.2%の減少。自費診療の提案を強化することを推奨します。',
      timestamp: new Date().toISOString(),
    },
    {
      id: 'alert-2',
      severity: 'error',
      title: '粗利率が目標を下回っています',
      message: '目標25%に対して24.7%。変動費の見直しを検討してください。',
      timestamp: new Date().toISOString(),
    },
  ];

  // モックデータ: 月次推移（直近6ヶ月）
  private mockTrends: MonthlyTrendData[] = [
    {
      yearMonth: '2024-05',
      totalRevenue: 9500000,
      operatingProfit: 2100000,
      newPatients: 55,
      returningPatients: 365,
      unitUtilization: 76.2,
      selfPayRate: 19.4,
    },
    {
      yearMonth: '2024-06',
      totalRevenue: 9100000,
      operatingProfit: 2020000,
      newPatients: 50,
      returningPatients: 355,
      unitUtilization: 75.8,
      selfPayRate: 19.1,
    },
    {
      yearMonth: '2024-07',
      totalRevenue: 8600000,
      operatingProfit: 1870000,
      newPatients: 46,
      returningPatients: 325,
      unitUtilization: 74.5,
      selfPayRate: 18.5,
    },
    {
      yearMonth: '2024-08',
      totalRevenue: 7500000,
      operatingProfit: 1560000,
      newPatients: 35,
      returningPatients: 285,
      unitUtilization: 71.2,
      selfPayRate: 17.8,
    },
    {
      yearMonth: '2024-09',
      totalRevenue: 8900000,
      operatingProfit: 1980000,
      newPatients: 49,
      returningPatients: 340,
      unitUtilization: 76.0,
      selfPayRate: 18.9,
    },
    {
      yearMonth: '2024-10',
      totalRevenue: 9300000,
      operatingProfit: 2170000,
      newPatients: 53,
      returningPatients: 358,
      unitUtilization: 78.5,
      selfPayRate: 18.2,
    },
  ];

  async getDashboardData(_clinicId: string): Promise<DashboardData> {
    // @MOCK_TO_API: GET /api/dashboard?clinic_id={_clinicId}
    // Request: clinicId (query param)
    // Response: DashboardData
    // Note: バックエンドで以下を実施:
    //   1. clinic_idに紐づく月次データを取得
    //   2. 主要KPIを自動計算（営業利益、粗利率、稼働率等）
    //   3. 前月比・前年比を算出
    //   4. アラート判定（目標未達、異常値検知）
    //   5. 直近6ヶ月の推移データを生成

    // モック実装: 200ms待機してデータ返却
    await new Promise((resolve) => setTimeout(resolve, 200));

    return {
      kpis: this.mockKpis,
      alerts: this.mockAlerts,
      trends: this.mockTrends,
      lastUpdated: '2025-10',
      dataSource: '手動入力',
    };
  }
}

File: backend/src/services/dashboard_service.py
from supabase import Client
from typing import List
from ..models.dashboard import DashboardData, DashboardKpi, DashboardAlert, MonthlyTrendData, KpiComparison
from datetime import datetime, timedelta
import uuid


class DashboardService:
    '''Dashboard service'''

    def __init__(self, supabase: Client):
        self.supabase = supabase

    def _calculate_trend(self, current: float, previous: float) -> str:
        '''Calculate trend direction'''
        if current > previous:
            return 'positive'
        elif current < previous:
            return 'negative'
        else:
            return 'neutral'

    def _calculate_percentage_change(self, current: float, previous: float) -> float:
        '''Calculate percentage change'''
        if previous == 0:
            return 0
        return ((current - previous) / previous) * 100

    async def get_dashboard_data(self, clinic_id: str) -> DashboardData:
        '''Get dashboard data for clinic'''
        try:
            # Get last 12 months of data
            response = self.supabase.table('monthly_data').select('*').eq('clinic_id', clinic_id).order('year_month', desc=True).limit(12).execute()

            monthly_data_list = response.data

            if not monthly_data_list or len(monthly_data_list) == 0:
                # Return empty dashboard
                return DashboardData(
                    kpis=[],
                    alerts=[],
                    trends=[],
                    last_updated=datetime.now().isoformat(),
                    data_source='No data available'
                )

            # Current month (most recent)
            current_month = monthly_data_list[0]

            # Previous month
            previous_month = monthly_data_list[1] if len(monthly_data_list) > 1 else None

            # Same month last year
            current_year_month = current_month['year_month']
            year, month = current_year_month.split('-')
            last_year_month = f'{int(year)-1}-{month}'

            last_year_data = None
            for data in monthly_data_list:
                if data['year_month'] == last_year_month:
                    last_year_data = data
                    break

            # Calculate KPIs
            kpis = self._calculate_kpis(current_month, previous_month, last_year_data)

            # Generate alerts
            alerts = self._generate_alerts(current_month, previous_month)

            # Generate trends
            trends = self._generate_trends(monthly_data_list)

            return DashboardData(
                kpis=kpis,
                alerts=alerts,
                trends=trends,
                last_updated=datetime.now().isoformat(),
                data_source='Monthly data input'
            )

        except Exception as e:
            raise ValueError(f'Failed to get dashboard data: {str(e)}')

    def _calculate_kpis(self, current: dict, previous: dict | None, last_year: dict | None) -> List[DashboardKpi]:
        '''Calculate KPIs'''
        kpis = []

        # Total Revenue
        current_revenue = current['total_revenue']
        prev_revenue = previous['total_revenue'] if previous else 0
        ly_revenue = last_year['total_revenue'] if last_year else 0

        kpis.append(DashboardKpi(
            id=str(uuid.uuid4()),
            label='Total Revenue',
            value=current_revenue,
            unit='¥',
            comparison=KpiComparison(
                trend=self._calculate_trend(current_revenue, prev_revenue),
                month_over_month=self._calculate_percentage_change(current_revenue, prev_revenue),
                year_over_year=self._calculate_percentage_change(current_revenue, ly_revenue)
            )
        ))

        # Operating Profit
        current_profit = current_revenue - (current['personnel_cost'] + current['material_cost'] + current['fixed_cost'] + current.get('other_cost', 0))
        prev_profit = (prev_revenue - (previous['personnel_cost'] + previous['material_cost'] + previous['fixed_cost'] + previous.get('other_cost', 0))) if previous else 0

        kpis.append(DashboardKpi(
            id=str(uuid.uuid4()),
            label='Operating Profit',
            value=current_profit,
            unit='¥',
            comparison=KpiComparison(
                trend=self._calculate_trend(current_profit, prev_profit),
                month_over_month=self._calculate_percentage_change(current_profit, prev_profit),
                year_over_year=0  # Simplified
            )
        ))

        # Total Patients
        current_patients = current['total_patients']
        prev_patients = previous['total_patients'] if previous else 0

        kpis.append(DashboardKpi(
            id=str(uuid.uuid4()),
            label='Total Patients',
            value=current_patients,
            unit='人',
            comparison=KpiComparison(
                trend=self._calculate_trend(current_patients, prev_patients),
                month_over_month=self._calculate_percentage_change(current_patients, prev_patients),
                year_over_year=0  # Simplified
            )
        ))

        return kpis

    def _generate_alerts(self, current: dict, previous: dict | None) -> List[DashboardAlert]:
        '''Generate alerts based on data'''
        alerts = []

        # Check for profit decline
        current_revenue = current['total_revenue']
        current_profit = current_revenue - (current['personnel_cost'] + current['material_cost'] + current['fixed_cost'] + current.get('other_cost', 0))

        if previous:
            prev_revenue = previous['total_revenue']
            prev_profit = prev_revenue - (previous['personnel_cost'] + previous['material_cost'] + previous['fixed_cost'] + previous.get('other_cost', 0))

            if current_profit < prev_profit:
                alerts.append(DashboardAlert(
                    id=str(uuid.uuid4()),
                    severity='warning',
                    title='Profit Decline',
                    message='Operating profit decreased compared to last month',
                    timestamp=datetime.now().isoformat()
                ))

            # Check for patient decline
            if current['total_patients'] < previous['total_patients']:
                alerts.append(DashboardAlert(
                    id=str(uuid.uuid4()),
                    severity='info',
                    title='Patient Count Decrease',
                    message='Total patients decreased compared to last month',
                    timestamp=datetime.now().isoformat()
                ))

        return alerts

    def _generate_trends(self, monthly_data_list: List[dict]) -> List[MonthlyTrendData]:
        '''Generate trend data'''
        trends = []

        for data in reversed(monthly_data_list):  # Oldest to newest
            total_revenue = data['total_revenue']
            operating_profit = total_revenue - (data['personnel_cost'] + data['material_cost'] + data['fixed_cost'] + data.get('other_cost', 0))

            # Calculate self-pay rate
            self_pay_rate = (data['self_pay_revenue'] / total_revenue * 100) if total_revenue > 0 else 0

            # Simplified unit utilization (placeholder calculation)
            unit_utilization = min(100, (data['treatment_count'] / 200 * 100)) if data['treatment_count'] > 0 else 0

            trends.append(MonthlyTrendData(
                year_month=data['year_month'],
                total_revenue=total_revenue,
                operating_profit=operating_profit,
                new_patients=data['new_patients'],
                returning_patients=data['returning_patients'],
                unit_utilization=unit_utilization,
                self_pay_rate=self_pay_rate
            ))

        return trends
